cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -2544,12 +2563,12 @@ private List<String> needsRestartBecauseAddedOrRemovedJbodVolumes(Pod pod, JbodS
                     Storage currentStorage = ModelUtils.decodeStorageFromJson(jsonStorage);
 
                     if (new StorageDiff(reconciliation, currentStorage, desiredStorage, currentReplicas, desiredReplicas).isVolumesAddedOrRemoved())    {
-                        return singletonList("JBOD volumes were added or removed");
+                        return RestartReasons.of(RestartReason.JBOD_VOLUMES_CHANGED);
                     }
                 }
             }
 
-            return new ArrayList<>();
+            return RestartReasons.empty();
```

Comment:
Maybe its only me. But what does empty RestartReason mean in terms of semantics? No restart? Or does it mean _restart for an unknown reason_? The list is a generic, so there it seemed to be more logical that empty list meant no restart. But with a speciific object, I would probably expect `null` to indicate no restart and `empty` RestartReason to indicate some kind of _unknown_ restart.

Reply:
Hi @scholzj I was copying the old semantics where "no reasons to restart" was an empty list. In fact, deep in the `KafkaRoller` there's an existing explicit null check.

`List<String> reasonToRestartPod = Objects.requireNonNull(podNeedsRestart.apply(pod));`

That said, I was thinking of something along your lines, where the code works on an `Optional<RestartReasons>` and an empty optional means "don't restart", whereas a non-empty optional, even if the RestartReasons pseudo-collection is empty, means, as you say - there is a restart occurring, but no explicit reasons were logged - that said, I haven't encountered that case yet. 

That way, I can add better logging around that last case if it occurs. How does that sounds?

So everywhere that was returning a `new ArrayList<>()/RestartReasons.empty()` would instead return an `Optional.empty()`. 

Alternatively, I could just change the existing RestartReasons class to express the same semantics - have a `RestartReasons.none()` method or similar, but I thought that Optional might be better. Keen on your thoughts? 

(I've always lived the trinary logic that `null` brings to the table :D)

Reply:
TBH, I personally think that Optional sucks. Old school null seems more practical. But I guess I can live with it if needed. I can for sure live with the empty list == no roll semantic. It just seems a bit weird when explicit object type is used.

Reply:
Yeah, I agree. For now I've renamed RestartReasons#isEmpty to RestartReasons#shouldRoll which returns true if the internal collection isn't empty. And I've change existing usages of that method to handle the inversion in logic.